src/java.base/share/classes/java/math/BigInteger.java

```diff
@@ -1844,6 +1862,86 @@ private static BigInteger multiplyKaratsuba(BigInteger x, BigInteger y) {
         }
     }
 
+    private abstract static sealed class RecursiveOp extends RecursiveTask<BigInteger> {
+        /**
+         * The threshold until when we should continue forking recursive ops
+         * if parallel is true. By default we use the Math.ceil() of
+         * log2(availableProcessors). Can be overridden with the system
+         * property -Djava.math.BigInteger.parallelForkThreshold=num. This
+         * threshold is only relevant for Toom Cook 3 multiply and square.
+         */
+        private static final int PARALLEL_FORK_THRESHOLD = Integer.getInteger(
+                "java.math.BigInteger.parallelForkThreshold",
+                (int) Math.ceil(Math.log(Runtime.getRuntime().availableProcessors()) / Math.log(2)));
+
+        @Serial
+        private static final long serialVersionUID = 0L;
+
+        private final boolean parallel;
+        private final int depth;
+
+        private RecursiveOp(boolean parallel, int depth) {
+            this.parallel = parallel;
+            this.depth = depth;
+        }
+
+        protected RecursiveTask<BigInteger> forkOrInvoke() {
+            if (parallel && depth <= PARALLEL_FORK_THRESHOLD) fork();
+            else invoke();
+            return this;
+        }
+
+        private static final class RecursiveMultiply extends RecursiveOp {
+            @Serial
+            private static final long serialVersionUID = 0L;
+
+            private final BigInteger a;
+            private final BigInteger b;
+
+            public RecursiveMultiply(BigInteger a, BigInteger b, boolean parallel, int depth) {
+                super(parallel, depth);
+                this.a = a;
+                this.b = b;
+            }
+
+            @Override
+            public BigInteger compute() {
+                return a.multiply(b, true, super.parallel, super.depth);
```

Comment:
I cannot remember ever using `super.` like this before either :-) I have a static inner class, which in turn has two static inner classes, which are its subclasses. The inner inner classes want to access the fields of their outer static super class. Since the fields are private, we need to use `super` otherwise we get this compiler error:

error: non-static variable parallel cannot be referenced from a static context
                return a.multiply(b, true, parallel, depth);

And if we move the inner inner class up one level, then we also get an error:

error: parallel has private access in RecursiveOp
            return a.multiply(b, true, parallel, super.depth);

A solution of course is to make the fields non-private. Then the compiler is happy and we can get rid of the `super`. Since the classes are private anyway, making the fields private does not increase the encapsulation. I will do that to get rid of the `super` FWIW.

Reply:
Oh, i did not realize that was a forced move.